File: backend/src/dtos/transaction_dto.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// ============ Transaction DTOs ============

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct CreateTransactionPayload {
    pub from_basket_id: i64,
    pub to_basket_id: i64,
    pub amount: f64,
    pub transaction_type_id: i64,
    pub title: String,
    pub description: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionResponse {
    pub id: i64,
    pub created_by_id: i64,
    pub from_basket_id: Option<i64>, // NULL for external deposits
    pub to_basket_id: i64,
    pub amount: f64,
    pub transaction_type_id: i64,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionWithDetails {
    pub id: i64,
    pub created_by_id: i64,
    pub from_basket: TransactionBasketInfo,
    pub to_basket: TransactionBasketInfo,
    pub amount: f64,
    pub transaction_type: TransactionTypeInfo,
    pub detail: Option<TransactionDetailResponse>,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionBasketInfo {
    pub id: i64,
    pub name: String,
    pub basket_type: String,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionTypeInfo {
    pub id: i64,
    pub name: String,
    pub parent_id: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionHistoryQuery {
    pub basket_id: Option<i64>,
    pub transaction_type_id: Option<i64>,
    pub from_date: Option<NaiveDateTime>,
    pub to_date: Option<NaiveDateTime>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

// ============ Transaction Type DTOs ============

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct CreateTransactionTypePayload {
    pub name: String,
    pub description: Option<String>,
    pub parent_id: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct UpdateTransactionTypePayload {
    pub name: Option<String>,
    pub description: Option<String>,
    pub parent_id: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionTypeChildrenResponse {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub name: String,
    pub description: Option<String>,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionTypeResponse {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub name: String,
    pub description: Option<String>,
    pub children: Vec<TransactionTypeChildrenResponse>,
    pub created_at: NaiveDateTime,
}

// ============ Transaction Detail DTOs ============

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct CreateTransactionDetailPayload {
    pub transaction_id: i64,
    pub title: String,
    pub description: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct UpdateTransactionDetailPayload {
    pub title: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct TransactionDetailResponse {
    pub id: i64,
    pub transaction_id: i64,
    pub title: String,
    pub description: Option<String>,
    pub created_at: NaiveDateTime,
}

File: frontend/app/dtos/transaction.dto.ts
import { z } from 'zod'
import { TransactionBasketInfoSchema } from './basket.dto'

export const TransactionTypeChildrenResponseSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable().optional(),
  parent_id: z.number().nullable().optional(),
  created_at: z.string()
})

export type TransactionTypeChildrenResponse = z.infer<typeof TransactionTypeChildrenResponseSchema>

export const TransactionTypeInfoSchema = z.object({
  id: z.number(),
  name: z.string(),
  parent_id: z.number().nullable().optional()
})

export type TransactionTypeInfo = z.infer<typeof TransactionTypeInfoSchema>

export const TransactionTypeResponseSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable().optional(),
  parent_id: z.number().nullable().optional(),
  children: z.array(TransactionTypeChildrenResponseSchema),
  created_at: z.string()
})

export type TransactionTypeResponse = z.infer<typeof TransactionTypeResponseSchema>

export const TransactionResponseSchema = z.object({
  id: z.number(),
  created_by_id: z.number(),
  from_basket_id: z.number().nullable().optional(),
  to_basket_id: z.number(),
  amount: z.number(),
  transaction_type_id: z.number(),
  created_at: z.string()
})

export type TransactionResponse = z.infer<typeof TransactionResponseSchema>

export const TransactionDetailResponseSchema = z.object({
  id: z.number(),
  transaction_id: z.number(),
  title: z.string(),
  description: z.string().nullable().optional(),
  created_at: z.string()
})

export type TransactionDetailResponse = z.infer<typeof TransactionDetailResponseSchema>

// Transaction With Details
export const TransactionWithDetailsSchema = z.object({
  id: z.number(),
  created_by_id: z.number(),
  from_basket: TransactionBasketInfoSchema,
  to_basket: TransactionBasketInfoSchema,
  amount: z.number(),
  transaction_type: TransactionTypeInfoSchema,
  detail: TransactionDetailResponseSchema.nullable().optional(),
  created_at: z.string()
})

export type TransactionWithDetails = z.infer<typeof TransactionWithDetailsSchema>

// Create Transaction Payload
export const CreateTransactionPayloadSchema = z.object({
  from_basket_id: z.number(),
  to_basket_id: z.number(),
  amount: z.number().min(0.01, 'Amount must be greater than 0'),
  transaction_type_id: z.number(),
  title: z.string().min(1, 'Title is required'),
  description: z.string().nullable().optional()
})

export type CreateTransactionPayload = z.infer<typeof CreateTransactionPayloadSchema>

// Create Transaction Type Payload
export const CreateTransactionTypePayloadSchema = z.object({
  name: z.string().min(1, 'Name is required'),
  description: z.string().nullable().optional(),
  parent_id: z.number().nullable().optional()
})

export type CreateTransactionTypePayload = z.infer<typeof CreateTransactionTypePayloadSchema>

// Update Transaction Type Payload
export const UpdateTransactionTypePayloadSchema = z.object({
  name: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  parent_id: z.number().nullable().optional()
})

export type UpdateTransactionTypePayload = z.infer<typeof UpdateTransactionTypePayloadSchema>
